### ISSUE TITLE
refactor idea: separate internal work from Quotes

### ISSUE THREAD

**Issue by marktoda** [Uniswap/uniswapx-parameterization-api#269]
One awkwardness in URA currently is that Quote objects (i.e. the main abstraction over responses) are conflated with internal work (i.e. the response from routing API, response from RFQ etc.). This causes several awkward pieces such as:
- `reparameterization` in DutchQuote to saturate an RFQ quote with classic data for the tail end of the auction
- portion data being inserted in several places after the fact, since fetching portion data is asynchronously done in routing API
- injecting trade size information and gas information for dutch quotes

Effectively, all "work" has been coupled with "quotes" which in reality they are not 1:1. The general purpose of this proposal is to decouple these things. My current thinking is as follows:

- Create a new `Task` interface, which represents any arbitrary independent task that should be run in parallel. Some examples of work:
  - fetch quote from routing API
  - fetch quote from RFQ
  - fetch portion from PortionService
  - fetch permit2 allowances from onchain
- `Task` interface should have some generic `run` function
- Change `Context` objects to depend on `Task` types instead of `Quote` types
- Keep the `Quote` interface, but have it only relate to creating the final API response from several task responses
  - i.e. DutchQuote can have a constructor that accepts `RFQResponse`, `ClassicResponse`, `PortionResponse`, and `Permit2Response`

Moving forward I think this is a much cleaner interface as, when we need to inject new pieces of data into our ever more complex quote types, we can just create new task runners and continue to parallelize and deduplicate work

**Comment by marktoda** [comment]
curious your thoughts @ConjunctiveNormalForm @zhongeric

**Comment by zhongeric** [comment]
I agree that the Quote entity feels very crowded lately, how do you envision a context to handle which async `tasks` must be resolved before the `Quote` can be made? The logic for DutchOrders right now includes quite a few conditionals already -  RFQ but no classic, classic and RFQ, open quote from classic, portion, etc. 

I'd imagine you can set timeouts on the tasks and force resolve/throw if certain resolving criteria are not met? Generally like this, just curious how this can simplify application logic for diff quote types
